playground/backend/internal/code_processing/code_processing.go

```diff
@@ -172,7 +172,11 @@ func compileStep(ctx context.Context, cacheService cache.Cache, paths *fs_tool.L
 			return nil
 		}
 	} else { // in case of Java, Go (not unit test), Scala - need compile step
-		executorBuilder := builder.Compiler(paths, sdkEnv)
+		executorBuilder, err := builder.Compiler(paths, sdkEnv)
```

Comment:
The fact that we see logging of an error and not returning it makes this method difficult to test.
The `compileStep` method signature should not have a `pipelineLifeCycleCtx context.Context` at the end. Typically context.Context is the first method argument.  I realize this is not part of the PR but it is making updates to this code more challenging.  Perhaps consider refactoring the method to do something like https://go.dev/play/p/f4fn1Scx1vE.

Reply:
There is a problem with this method that it takes two contexts, this is something which will require careful consideration on why this was done this way and how it can be rectified.
Also refactoring the call chain leading to this function so it will be able to return the error also will require some thought and it goes beyond the scope of this PR. 
Filed a task to resolve this separately: https://github.com/apache/beam/issues/25277